Add bookings

Believerin opened the pull request.

Believerin committed:
Initial commit

Believerin committed:
Initial commit

Believerin committed:
Initial commit

Believerin committed:
perf: "проверка на дублирование почты изменена и перенесена в UserDbStorage"
fix: "в метод getItem добавлена обработка null"
feat: "в dto добавлены аннотации валидации"
feat: "аннотация @Component заменёна на @Repository"

Believerin committed:
feat: "добавлена функция бронирования и комментирования"

Believerin committed:
feat: "добавлена функция бронирования и комментирования"

TyutterinYakov commented on pom.xml:
```diff
@@ -54,6 +55,31 @@
+			<version>3.7.0</version>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.data</groupId>
```
Точно ли требуется эта зависимость?)

Believerin replied:
Убрал

TyutterinYakov commented on pom.xml:
```diff
@@ -54,6 +55,31 @@
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.data</groupId>
+			<artifactId>spring-data-rest-webmvc</artifactId>
```
Точно ли требуется эта зависимость?)

Believerin replied:
Убрал

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,52 @@
+import javax.validation.Valid;
+import java.util.Collection;
 
 /**
```
Так как функционал класса реализован - комментарий более не требуется) Аналогично в других местах

Believerin replied:
Убрал

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,52 @@
  * TODO Sprint add-bookings.
  */
 @RestController
+@AllArgsConstructor
```
В данном случае следует использовать `@RequiredArgsConstructor`, так как `@AllArgsConstructor` используется для объявления конструктора для всех полей, а нам требуется только для финальных)

Believerin replied:
Удаление `@AllArgsConstructor` приводит к невозможности работы `@Builder`, приходится генерировать конструктор со всеми полями в любом случае.

TyutterinYakov replied:
А для чего билдер для класса-контроллера?)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/exception/ErrorResponse.java:
```diff
@@ -0,0 +1,14 @@
+
+class ErrorResponse {
+
+    String error;
```
Для соблюдения инкапсуляции поля класса должны быть приватными)

Believerin replied:
Установил `private`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,55 @@
+
+    private final ItemService itemService;
+
+    public ItemController(ItemService itemService) {
```
Так же можно воспользоваться аннотацией `@RequiredArgsConstructor` из lombok, благодаря которой автоматически будет сгенерирован конструктор для финальных полей)

Believerin replied:
Установил

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/ItemMapper.java:
```diff
@@ -0,0 +1,37 @@
+
+import ru.practicum.shareit.item.dto.ItemDto;
+import ru.practicum.shareit.item.model.Item;
+
```
(предложение)Можно добавить такую аннотацию
`@NoArgsConstructor(access = AccessLevel.PRIVATE)` или [`@UtilityClass`](https://projectlombok.org/features/experimental/UtilityClass)
Чтобы создать приватный конструктор без параметров, тем самым запретив создание объектов этого класса) Так это не имеет смысла ввиду того, что все в этом классе static, а значит класс не имеет состояния)

Believerin replied:
Установил

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/ItemMapper.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    public static Item toItem(int userId, Item modifyingItem, ItemDto itemDto) {
+        return Item.builder()
```
Эту логику следует вынести в сервис, так как это часть бизнес-логики. Маппер занимается только лишь преобразованиями одного типа к другому, иначе нарушается принцип единственной ответственности

Believerin replied:
Вынес в сервис

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/ItemMapper.java:
```diff
@@ -0,0 +1,37 @@
+    public static Item toItem(int userId, Item modifyingItem, ItemDto itemDto) {
+        return Item.builder()
+                .id(itemDto.getId() != null ? itemDto.getId() : modifyingItem.getId())
+                .name(itemDto.getName() != null ? itemDto.getName() : modifyingItem.getName())
```
Хорошо было бы также проверить, что строка не пуста и не состоит только из пробелов) Сделать это удобно с помощью метода `isBlank`)

Believerin replied:
Сделал проверку

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/ItemMapper.java:
```diff
@@ -0,0 +1,37 @@
+        return Item.builder()
+                .id(itemDto.getId() != null ? itemDto.getId() : modifyingItem.getId())
+                .name(itemDto.getName() != null ? itemDto.getName() : modifyingItem.getName())
+                .description(itemDto.getDescription() != null ? itemDto.getDescription() : modifyingItem.getDescription())
```
Хорошо было бы также проверить, что строка не пуста и не состоит только из пробелов) Сделать это удобно с помощью метода `isBlank`)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/UserMapper.java:
```diff
@@ -0,0 +1,30 @@
+        );
+    }
+
+    public static User toUser(int userId, User modifyingUser, UserDto userDto) {
```
Аналогично, эта логика должна находиться в сервисе

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/UserMapper.java:
```diff
@@ -0,0 +1,30 @@
+    public static User toUser(int userId, User modifyingUser, UserDto userDto) {
+        return new User(
+                userId,
+                userDto.getName() != null ? userDto.getName() : modifyingUser.getName(),
```
Хорошо было бы также проверить, что строка не пуста и не состоит только из пробелов) Сделать это удобно с помощью метода `isBlank`)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/UserMapper.java:
```diff
@@ -0,0 +1,30 @@
+        return new User(
+                userId,
+                userDto.getName() != null ? userDto.getName() : modifyingUser.getName(),
+                userDto.getEmail() != null ? userDto.getEmail() : modifyingUser.getEmail()
```
Хорошо было бы также проверить, что строка не пуста и не состоит только из пробелов) Сделать это удобно с помощью метода `isBlank`)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/UserMapper.java:
```diff
@@ -0,0 +1,30 @@
+
+import ru.practicum.shareit.user.dto.UserDto;
+import ru.practicum.shareit.user.model.User;
+
```
(предложение)Можно добавить такую аннотацию
`@NoArgsConstructor(access = AccessLevel.PRIVATE)` или [`@UtilityClass`](https://projectlombok.org/features/experimental/UtilityClass)
Чтобы создать приватный конструктор без параметров, тем самым запретив создание объектов этого класса) Так это не имеет смысла ввиду того, что все в этом классе static, а значит класс не имеет состояния)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,45 @@
+    }
+
+    @PostMapping
+    public UserDto addUser(@Valid @RequestBody UserDto userDto) {
```
Кстати, вместо @Valid можно использовать такую аннотацию @Validated(Create.class). Таким образом мы указываем, какую группу аннотаций следует учитывать. Create - это интерфейс, который ты создаешь самостоятельно, он будет пустым и служит лишь для указания группы) Аналогично следует поступить при обновлении, но там будет аннотация @Validated(Update.class)

Believerin replied:
Да, я делал, но потом понял, что она не задействуется в тестах

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/UserController.java:
```diff
@@ -1,12 +1,45 @@
+    }
+
+    @PatchMapping("{userId}")
+    public UserDto modifyUser(@PathVariable int userId, @RequestBody UserDto userDto) {
```
Уточнение в виде "user", "users" и тд не требуется, так как методы не летают в воздухе, а находятся в контексте определенных классов) Аналогично следует поступить с другими методами этого класса и вообще со всеми методами в проекте)

Believerin replied:
Убрал лишнее

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/dto/UserDto.java:
```diff
@@ -0,0 +1,19 @@
+@Setter
+@EqualsAndHashCode
+@AllArgsConstructor
+public class UserDto {
```
Так же и здесь, можно было бы использовать валидацию по группам)
Для имени:
`@NotBlank(groups = {Create.class})`
Для почты аннотацию `@Email` следует учитываться для обновления и создания, так как, если она не будет передана вообще, то все будет окей, а если будет, тогда она будет проверена на соответствие почте, а вот аннотацию
`@NotBlank` только для создания) Так как при обновлении передавать ее не обязательно)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -0,0 +1,30 @@
+    @NotBlank
+    private String name;
+    @Email
+    @NotNull
```
Аннотация `@NotNull` не требуется, так как проверка на null входит в `@NotBlank`

Believerin replied:
Исправил

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -0,0 +1,30 @@
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private int id;
+    @NotNull
```
Аннотация `@NotNull` не требуется, так как проверка на null входит в `@NotBlank`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -0,0 +1,30 @@
+@Entity
+@Getter
+@Setter
+@EqualsAndHashCode
```
Перед использованием аннотаций автоматической генерации кода над классами-сущностями, хорошо было бы посмотреть статью
https://habr.com/ru/company/haulmont/blog/564682/

TyutterinYakov commented on src/main/java/ru/practicum/shareit/request/ItemRequest.java:
```diff
@@ -1,7 +1,26 @@
+@Entity
+@Getter
+@Setter
+@EqualsAndHashCode
```
Перед использованием аннотаций автоматической генерации кода над классами-сущностями, хорошо было бы посмотреть статью
https://habr.com/ru/company/haulmont/blog/564682/

Believerin replied:
Выставил `@EqualsAndHashCode.Exclude`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,7 +1,35 @@
+@Entity
+@Getter
+@Setter
+@EqualsAndHashCode
```
Перед использованием аннотаций автоматической генерации кода над классами-сущностями, хорошо было бы посмотреть статью
https://habr.com/ru/company/haulmont/blog/564682/

TyutterinYakov commented on src/main/java/ru/practicum/shareit/comment/model/Comment.java:
```diff
@@ -0,0 +1,31 @@
+@Entity
+@Getter
+@Setter
+@EqualsAndHashCode
```
Перед использованием аннотаций автоматической генерации кода над классами-сущностями, хорошо было бы посмотреть статью
https://habr.com/ru/company/haulmont/blog/564682/

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -0,0 +1,38 @@
+@Entity
+@Getter
+@Setter
+@EqualsAndHashCode
```
Перед использованием аннотаций автоматической генерации кода над классами-сущностями, хорошо было бы посмотреть статью
https://habr.com/ru/company/haulmont/blog/564682/

TyutterinYakov commented on src/main/java/ru/practicum/shareit/comment/CommentMapper.java:
```diff
@@ -0,0 +1,30 @@
+import ru.practicum.shareit.user.model.User;
+
+import java.time.LocalDateTime;
+
```
(предложение)Можно добавить такую аннотацию
`@NoArgsConstructor(access = AccessLevel.PRIVATE)` или [`@UtilityClass`](https://projectlombok.org/features/experimental/UtilityClass)
Чтобы создать приватный конструктор без параметров, тем самым запретив создание объектов этого класса) Так это не имеет смысла ввиду того, что все в этом классе static, а значит класс не имеет состояния)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/BookingMapper.java:
```diff
@@ -0,0 +1,31 @@
+import ru.practicum.shareit.booking.model.Booking;
+import ru.practicum.shareit.item.model.Item;
+import ru.practicum.shareit.user.model.User;
+
```
(предложение)Можно добавить такую аннотацию
`@NoArgsConstructor(access = AccessLevel.PRIVATE)` или [`@UtilityClass`](https://projectlombok.org/features/experimental/UtilityClass)
Чтобы создать приватный конструктор без параметров, тем самым запретив создание объектов этого класса) Так это не имеет смысла ввиду того, что все в этом классе static, а значит класс не имеет состояния)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,52 @@
+
+    @GetMapping
+    public Collection<BookingDto> getAllBookingsByBooker(@RequestHeader("X-Sharer-User-Id") int bookerId,
+                                                         @RequestParam(required = false, defaultValue = "ALL")
```
Если указано `defaultValue`, это уже подразумевает, что параметр не является обязательным, следовательно `required = false` не требуется)

Believerin replied:
Убрал

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/BookingController.java:
```diff
@@ -1,12 +1,52 @@
+
+    @GetMapping("owner")
+    public Collection<BookingDto> getAllBookingsByOwner(@RequestHeader("X-Sharer-User-Id") int ownerId,
+                                                         @RequestParam(required = false, defaultValue = "ALL")
```
Если указано `defaultValue`, это уже подразумевает, что параметр не является обязательным, следовательно `required = false` не требуется)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDto.java:
```diff
@@ -1,7 +1,40 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
 public class BookingDto {
```
Хорошо было бы разбить DTO на несколько, чтобы один выступал в качестве тела запроса, а другой в качестве тела ответа, чтобы не производить передачу лишних данных)

TyutterinYakov replied:
Хорошо было бы разбить DTO на несколько, чтобы один выступал в качестве тела запроса, а другой в качестве тела ответа, чтобы не производить передачу лишних данных)

TyutterinYakov replied:
Хорошо было бы разбить DTO на несколько, чтобы один выступал в качестве тела запроса, а другой в качестве тела ответа, чтобы не производить передачу лишних данных)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/comment/dto/CommentDto.java:
```diff
@@ -0,0 +1,31 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@Builder
+public class CommentDto {
```
Хорошо было бы разбить DTO на несколько, чтобы один выступал в качестве тела запроса, а другой в качестве тела ответа, чтобы не производить передачу лишних данных)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,29 @@
+@EqualsAndHashCode
+@AllArgsConstructor
+@Builder
 public class ItemDto {
```
Хорошо было бы разбить DTO на несколько, чтобы один выступал в качестве тела запроса, а другой в качестве тела ответа, чтобы не производить передачу лишних данных)

TyutterinYakov replied:
Хорошо было бы разбить DTO на несколько, чтобы один выступал в качестве тела запроса, а другой в качестве тела ответа, чтобы не производить передачу лишних данных)

TyutterinYakov replied:
Хорошо было бы разбить DTO на несколько, чтобы один выступал в качестве тела запроса, а другой в качестве тела ответа, чтобы не производить передачу лишних данных)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/exception/ErrorHandler.java:
```diff
@@ -0,0 +1,45 @@
+import org.springframework.web.bind.annotation.*;
+
+@RestControllerAdvice
+public class ErrorHandler {
```
В класс следует добавить логирование) Для этого достаточно использовать аннотацию из lombok ->> [@slf4j](https://github.com/slf4j). Далее в каждый наш хендлер следует добавить логирование сообщения об ошибке)
Так же следует добавить хендлер для обработки всех необработанных исключений, ловить будем `Throwable`, а возвращать статус 500 INTERNAL_SERVER_ERROR

Believerin replied:
Это исключение не обрабатывается, т.к. оно в тестах не выбрасывается

Believerin replied:
Сделал

TyutterinYakov replied:
Отсутствует логгирование сообщений)

Believerin replied:
Я добавил @Slf4j, она там точно есть

TyutterinYakov commented on src/main/java/ru/practicum/shareit/comment/dto/CommentDto.java:
```diff
@@ -0,0 +1,31 @@
+    private String text;
+    private int item;
+    private String authorName;
+    private User author;
```
В классах-DTO не должно содержаться классов-моделей, мы всегда должны контролировать те данные, что мы получаем и отдаем)

Believerin replied:
Убрал сущности из  БД и заменил на DTO

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDto.java:
```diff
@@ -1,7 +1,40 @@
+    @Future
+    private LocalDateTime end;
+    private Integer itemId;
+    private Item item;
```
В классах-DTO не должно содержаться классов-моделей, мы всегда должны контролировать те данные, что мы получаем и отдаем)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDto.java:
```diff
@@ -1,7 +1,40 @@
+    private Integer itemId;
+    private Item item;
+    private Integer bookerId;
+    private User booker;
```
В классах-DTO не должно содержаться классов-моделей, мы всегда должны контролировать те данные, что мы получаем и отдаем)
Удобно было бы представить их в виде двух внутренних классов)
```
	@Data
	public static class Booker {
		private final long id;
		private final String name;
	}

	@Data
	public static class Item {
		private final long id;
		private final String name;
	}
```

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDto.java:
```diff
@@ -1,7 +1,40 @@
 public class BookingDto {
-}
+    private Integer id;
+    @Future
```
`@FutureOrPresent`, так как дата начала бронирования может быть равна текущей, либо же быть из будущего. Не забываем о `@NotNull`, так как первая аннотация допускает null)

Believerin replied:
Установил

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDto.java:
```diff
@@ -1,7 +1,40 @@
+    private Integer id;
+    @Future
+    private LocalDateTime start;
+    @Future
```
Не забываем использовать `@NotNull`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDto.java:
```diff
@@ -1,7 +1,40 @@
+    private LocalDateTime start;
+    @Future
+    private LocalDateTime end;
+    private Integer itemId;
```
Идентификатор так же должен быть передан в любом случае, следует использовать `@NotNull`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -0,0 +1,38 @@
+public class Booking {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Integer id;
```
Хорошо было в качестве типа для идентификатора использовать тип `Long`, так как при большом количестве операций вставки - типа `Integer` на долго не хватит)
Аналогично с другими сущностями)

Believerin replied:
Да, согласен, но для этого задания можно и `Integer` обойтись

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+        try {
+            User user = userRepository.save(UserMapper.toUser(userDto));
+            return UserMapper.toUserDto(user);
+        } catch (EmptyResultDataAccessException e) {
```
Почему бы н добавить еще один хендлер для этого исключения - `EmptyResultDataAccessException` и возвращать требуемый статус после обработки?)

Believerin replied:
Добавил

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+        try {
+            User user = UserMapper.toUser(userId, modifyingUser, userDto);
+            return UserMapper.toUserDto(userRepository.save(user));
+        } catch (EmptyResultDataAccessException e) {
```
Почему бы н добавить еще один хендлер для этого исключения - `EmptyResultDataAccessException` и возвращать требуемый статус после обработки?)

Believerin replied:
Добавил

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+import java.util.*;
+
+@Service
+@RequiredArgsConstructor
```
Так как ты познакомился с транзакциями, хорошо было бы все сервисы сделать транзакционными)
Для этого необходимо навесить над ними аннотацию 
`@Transactional(readOnly = true)`, что указывает на то, что фиксация изменений не будет производиться) А над теми методами, где требуется фиксация в БД, необходимо использовать `@Transactional`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public UserDto updateUser(int userId, UserDto userDto) {
+        User modifyingUser;
+        Optional<User> o = userRepository.findById(userId);
```
В данном случае удобно было бы использовать `orElseThrow(() -> ...)`
Чтобы запись была более краткой)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+        }
+        try {
+            User user = UserMapper.toUser(userId, modifyingUser, userDto);
+            return UserMapper.toUserDto(userRepository.save(user));
```
Вызов метода save более не будет требоваться, так как мы будем работать в рамках транзакции с полученным объектом класса-сущности из БД, а значит изменения будут автоматически зафиксированы)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/UserRepository.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface UserRepository extends JpaRepository<User, Integer> {
+
+    List<UserDto> findAllBy(Class<UserDto> type);
```
Более корректно было бы работать здесь с классом-сущностью, так как c DTO более корректно работать на уровне контроллера или сервиса, не доходя до уровня доступа к данным

TyutterinYakov commented on src/main/java/ru/practicum/shareit/comment/CommentRepository.java:
```diff
@@ -0,0 +1,16 @@
+            "c.id, c.text, c.item, c.authorName, c.created) " +
+            "from Comment as c " +
+            "where c.item = ?1")
+    List<CommentDto> findAllByItem(int itemId, Class<CommentDto> type);
```
Более корректно было бы работать здесь с классом-сущностью, так как c DTO более корректно работать на уровне контроллера или сервиса, не доходя до уровня доступа к данным

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+
+    @Override
+    public BookingDto addBooking(BookingDto bookingDto, int bookerId) {
+        if (bookingDto.getEnd().isBefore(bookingDto.getStart())) {
```
Не забываем проверить, что время начала бронирования не равно времени окончания)
Если будет интересно, то эту валидацию можно было бы реализовать с помощью аннотации валидации над классом, чтобы поддерживать декларативный подход
https://devcolibri.com/spring-mvc-кастомная-аннотация-для-валидации/
Кстати, если будет интересно то, как реализовать кастомную аннотацию валидации на уровне класса `BookingRequestDto` - можешь чекнуть тут) https://github.com/TyutterinYakov/CustomValidationClassLevel По примеру может быть легче разобраться с тем, что требуется)

Believerin replied:
Добавил

Believerin replied:
За статьи спасибо, почитаю

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/BookingRepository.java:
```diff
@@ -0,0 +1,49 @@
+
+public interface BookingRepository extends JpaRepository<Booking, Integer> {
+
+    @Query(value = "select new ru.practicum.shareit.booking.dto.BookingDto(" +
```
Аналогично, следует изменить методы таким образом, чтобы работа осуществлялась с классами-сущностями

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,139 @@
+
+    @Override
+    public Collection<ItemDto> findAllOwn(int userId) {
+        return itemRepository.findByOwner(userId).stream()
```
Следует получить комментарии всех вещей, которые будут возвращены вне цикла и разбить их на мапу, где ключ - идентификатор вещи, а значение - коллекция из комментариев) Таким образом мы не будем производить n-лишних запросов к БД) Получение следует произвести вне цикла)
Аналогично следует поступить и с поиском предыдущего и следующего бронирования)
Где можно было бы по коллекции предметов произвести получение и сразу же выполнить сортировку, чтобы поиск небходимых бронирований был проще)
```
		Map<Item, List<Comment>> comments = commentRepository.findByItemIn(items, Sort.by(DESC, "created"))
				.stream()
				.collect(groupingBy(Comment::getItem, toList()));
```
Это лишь предложение по реализации, можно реализовать свою логику, главное избавить об обращения к репозиториям в циклах)

Believerin replied:
Сделал без лишних запросов в БД

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+        }
+        Booking b = BookingMapper.toBooking(bookingDto, item, booker);
+        b.setStatus(BookingStatus.WAITING);
+        Booking booking = bookingRepository.save(b);
```
Вызов метода save более не будет требоваться, так как мы будем работать в рамках транзакции с полученным объектом класса-сущности из БД, а значит изменения будут автоматически зафиксированы)

Believerin replied:
Тут сохраняется новый объект, поэтому оставлю строку.

Believerin replied:
В `acceptOrDeny` я испольpую кастомный метод c `@Query` и там транзакция не проходит

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,107 @@
+            throw new ValidationException("начало периода позже, чем его конец");
+        }
+        Item item;
+        Optional<Item> o = itemRepository.findById(bookingDto.getItemId());
```
При именовании переменных следует использовать более осмысленные имена, так как это может повлиять на читаемость и понимание кода)
В данном случае следует использовать `orElseThrow(() -> ...)`
Чтобы запись была более краткой) Аналогично в других местах

TyutterinYakov reviewed (commented):
Добрый вечер! В целом - хорошо :) Отметил то, что следует исправить на данном этапе и то, что можно сделать лучше! Хорошего вечера) Если возникнут вопросы - пиши в пачку, буду рад помочь🌃

Believerin committed:
perf: "Убраны лишние обращения к БД"
refactor: "Добавлены аннотации @Transactional"
refactor: "Методы вынесены из маппера в сервис"

Believerin committed:
perf: "Убраны лишние обращения к БД"
refactor: "Добавлены аннотации @Transactional"
refactor: "Методы вынесены из маппера в сервис"

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -0,0 +1,41 @@
+    private LocalDateTime start;
+    @Column(name = "end_date")
+    private LocalDateTime end;
+    @ToString.Exclude
```
Но выше мы работаем не с `@ToString`, а с `@EqualsAndHashCode`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -0,0 +1,41 @@
+import javax.persistence.*;
+import java.time.LocalDateTime;
+
+/**
```
Не забываем убрать комментарии)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDto.java:
```diff
@@ -1,7 +1,44 @@
+import javax.validation.constraints.*;
+import java.time.LocalDateTime;
+
 /**
```
Комментарии следует убрать везде, где реализован функционал

TyutterinYakov commented on src/main/java/ru/practicum/shareit/comment/CommentRepository.java:
```diff
@@ -0,0 +1,18 @@
+            "c.id, c.text, c.item, c.authorName, c.created) " +
+            "from Comment as c " +
+            "where c.item = ?1")
+    List<Comment> findAllByItem(int itemId);
```
Можно было бы не прибегать к `@Query` для простых запросов)
`List<Comment> findAllByItem_id(int itemId);`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,113 @@
+    public Collection<BookingDto> getAllByBooker(int bookerId, String state) {
+        userService.get(bookerId);
+        int status;
+        try {
```
Почему бы не вынести эту логику в отдельный приватный метод, чтобы не дублировать код?)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/service/BookingServiceImpl.java:
```diff
@@ -0,0 +1,113 @@
+    public Collection<BookingDto> getAllByOwner(int ownerId, String state) {
+        userService.get(ownerId);
+        int status;
+        try {
```
Почему бы не вынести эту логику в отдельный приватный метод, чтобы не дублировать код?)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/comment/model/Comment.java:
```diff
@@ -0,0 +1,33 @@
+    private String text;
+    @Column(name = "item_id")
+    private int item;
+    @ToString.Exclude
```
Разве `@ToString.Exclude`?)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,31 @@
+import javax.validation.constraints.*;
+import java.util.List;
+
 /**
```
Комментарии следует исключить везде, где реализован функционал

TyutterinYakov replied:
Комментарии следует исключить везде, где реализован функционал

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+            return UserMapper.toUserDto(o.get());
+        } else {
+            throw new NoSuchBodyException("Запрашиваемый пользователь");
+        }
```
Почему бы не использовать более короткую запись?)
```
User user = userRepository.findById(id).orElseThrow(() -> new NoSuchBodyException("Запрашиваемый пользователь"));
return UserMapper.toUserDto(user);
```

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+
+    //--------------------------------------Служебный метод-------------------------------------------------
+
+    public static User toUser(int userId, User updatingUser, UserDto userDto) {
```
Так как метод используется для внутренней реализации - следует сделать его приватным

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+
+    public static User toUser(int userId, User updatingUser, UserDto userDto) {
+        updatingUser.setId(userId);
+        if (userDto.getName() != null && !userDto.getName().isBlank()) {
```
Выброс исключения и второе условие не требуется:
```
if (userDto.getName() != null && !userDto.getName().isBlank()) {
        updatingUser.setName(userDto.getName());
}
```
Так как при обновлении нам важно понять лишь одно - производить изменение значения или нет)
Аналогичная проверка должна быть реализована и для почты, так как `@Email` допускает пустую строку и null - за корректное значение)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+
+    private final UserRepository userRepository;
+    @Autowired
+    private TransactionTemplate template;
```
Он здесь точно требуется?)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,189 @@
+            userService.get(userId);
+            Item item = itemRepository.save(ItemMapper.toItem(userId, itemDto));
+            return ItemMapper.toItemDto(item);
+        } catch (NoSuchBodyException e) {
```
Зачем мы сначала ловим исключение, а затем повторно его выбрасываем?)

Believerin replied:
Потому что в его обработчике есть общая фраза - "отсутствует или не найден", а в отлавливании конкретное сообщение по сущности

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,189 @@
+                    .collect(Collectors.toList());
+                    itemDto.setComments(comments);
+                }).collect(Collectors.toList());
+        return itemsWithComments.stream().peek(itemDto -> {
```
Метод `getLastAndNextBookings` вызывается в цикле, а он, в свою очередь, вызывает метод репозитория `findByItemIdInOrderByStartAsc`, из-за чего плодятся n-лишние запросы к БД. Следует этого так же избежать)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,189 @@
+
+    private Map<String, BookingDto> getLastAndNextBooking(int itemId) {
+        Map<String, BookingDto> nearestBookings = new HashMap<>();
+        List<Booking> bookings = bookingRepository.findByItemIdOrderByStartAsc(itemId);
```
Получать следует только бронирования в состоянии `APPROVED`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,189 @@
+
+    private Map<Integer, Map<String, BookingDto>> getLastAndNextBookings(List<Integer> itemIds) {
+        Map<Integer, Map<String, BookingDto>> nearestBookingsByItemId = new HashMap<>();
+        List<Booking> bookingsByItemIds = bookingRepository.findByItemIdInOrderByStartAsc(itemIds);
```
Получать следует только бронирования в состоянии `APPROVED`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,189 @@
+                .orElse(null);
+        Booking last = bookings.stream()
+                .sorted(Comparator.comparing(Booking::getStart).reversed())
+                .filter(booking -> booking.getStart().isBefore(LocalDateTime.now()))
```
Так же следует учитывать то, что, если время старта равно текущему, бронирование так же является последним(текущим)

TyutterinYakov reviewed (commented):
Добрый вечер! Отметил оставшиеся недочеты :) Хорошего вечера! Если возникнут вопросы - пиши в пачку, буду рад помочь!🌃 Кстати, если есть выбор, то лучшим решением было бы написать более простые запросы для получения бронирований в разных состояниях, а не писать пару сложны запросов, это более лучшее решение со стороны производительности и простоты кода)

Believerin committed:
perf: "Убраны лишние обращения к БД"
refactor: "Убрано возвращение лишних выходных данных"
refactor: "Удалено дублирование кода извлечением в метод"

TyutterinYakov commented on src/main/java/ru/practicum/shareit/comment/CommentRepository.java:
```diff
@@ -0,0 +1,18 @@
+
+public interface CommentRepository extends JpaRepository<Comment, Integer> {
+
+/*    @Query(value = "select new ru.practicum.shareit.comment.model.Comment(" +
```
Если код не используется - следует его исключить

TyutterinYakov reviewed (commented):
Добрый день! Отметил оставшиеся недочеты :) Хорошего дня! Если возникнут вопросы - пиши в пачку, буду рад помочь🌅

Believerin committed:
feature: "Добавлены классы ResponseDto И RequestDto"
feature: "Добавлены классы для полученияи возвращения Item"
feature: "Добавлены классы для полученияи возвращения Booking"

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDto.java:
```diff
@@ -1,7 +1,42 @@
+@NoArgsConstructor
+@Builder
 public class BookingDto {
-}
```
Действительно ли в этом классе требуются аннотации валидации?) Если я не ошибаюсь, то он выступает только в качестве возвращаемого значения

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/dto/ItemDto.java:
```diff
@@ -1,7 +1,34 @@
+@NoArgsConstructor
+@Builder
 public class ItemDto {
-}
```
Действительно ли в этом классе требуются аннотации валидации?) Если я не ошибаюсь, то он выступает только в качестве возвращаемого значения

TyutterinYakov reviewed (approved):
Добрый вечер! Поздравляю с успешной сдачей спринта :) Сил и терпения на этом сложном, но интересном пути! Успехов в обучении и прекрасного вечера🌃🥳
